### ISSUE TITLE
Using a @deprecated named parameter should not be an Analyzer warning inside of the method that defines it.

### ISSUE THREAD

**Issue by jonahwilliams** [dart-lang/sdk#32468]
I would like to give the users of a Flutter API warning about the removal of a named parameter.  If I mark it deprecated, however, this becomes an Analyzer warning inside of that same method.

Assuming that the point of deprecated is to allow a transition period, it doesn't really make sense to warn inside of this method.  I need to keep using that parameter if i don't want to break my API all at once, and if I wanted to break my API all at once I wouldn't bother with @deprecated.

As I workaround I am using an //ignore to silence the lint.

Example:
```dart
Future<T> showDialog<T>({
  @required BuildContext context,
  bool barrierDismissible: true,
  @deprecated Widget child,
  WidgetBuilder builder,
}) {
  assert(child == null || builder == null); // <- this is a warning
  return Navigator.of(context, rootNavigator: true).push(new _DialogRoute<T>(
    child: child ?? new Builder(builder: builder), // <- so is this
    theme: Theme.of(context, shadowThemeOnly: true),
    barrierDismissible: barrierDismissible,
    barrierLabel: MaterialLocalizations.of(context).modalBarrierDismissLabel,
  ));
}


``

**Comment by bwilkerson** [comment]
https://dart-review.googlesource.com/c/sdk/+/45840

**Comment by jonahwilliams** [comment]
Thank you for the quick fix!

**Comment by bwilkerson** [comment]
You're welcome. I'm just sorry we didn't think of that case before you ran into it.

**Comment by shilangyu** [comment]
This now seems to not give a warning on the calling site! Given the same example from flutter with `showDialog` where the `child` argument is marked as deprecated: no warnings are shown when I pass a `child` argument to `showDialog`! We discovered our codebase used this exact argument without ever knowing it is deprecated (`flutter analyze` nor vscode showed any warnings) and realized it only once our builds broke when upgrading flutter. Should I open a separate issue for that?

**Comment by bwilkerson** [comment]
@srawlins @scheglov

**Comment by scheglov** [comment]
Hm... Cannot reproduce, at least with older Flutter version.
![image](https://user-images.githubusercontent.com/384794/105416528-25895280-5bef-11eb-8b0e-314589768931.png)
Also cannot reproduce with the most recent Flutter form git.

**Comment by shilangyu** [comment]
VScode with flutter stable:

![image](https://user-images.githubusercontent.com/29288116/105464751-ee1bb400-5c89-11eb-919f-26c218a7382e.png)

No warnings in vscode nor after running `flutter analyze`.

Here is an example of a different deprecated property that correctly is detected as such:

![image](https://user-images.githubusercontent.com/29288116/105465027-536fa500-5c8a-11eb-8ad0-0b10398cd3fb.png)

**Comment by scheglov** [comment]
Sorry, I still cannot reproduce it :-(

**Comment by shilangyu** [comment]
I reproduced it on my linux and windows machines. My friend also can reproduced it on MacOs.

In fact we run `flutter analyze` with fatal warnings in the CI, so would know if it was detected. Thus this also is reproduced in an isolated CI environment. I don't know how else I can show that there are no deprecation warnings.

**Comment by scheglov** [comment]
I can reproduce this issue in `stable` channel, but it looks fixed in `beta` channel.
All the time before I was trying to reproduce on the `master` channel.

So, I think the issue is fixed, but not rolled into stable yet.

```console
scheglov@scheglov-macbookpro2:~/Source/flutter2/flutter/examples/hello_world (stable)$
scheglov@scheglov-macbookpro2:~/Source/flutter2/flutter/examples/hello_world (stable)$ /Users/scheglov/Source/flutter2/flutter/bin/flutter analyze
Analyzing hello_world...

warning • The parameter 'context' is required • lib/main.dart:9:3 • missing_required_param

1 issue found. (ran in 10.5s)
scheglov@scheglov-macbookpro2:~/Source/flutter2/flutter/examples/hello_world (stable)$ /Users/scheglov/Source/flutter2/flutter/bin/flutter channel beta
Switching to flutter channel 'beta'...
git: Switched to a new branch 'beta'
git: M	examples/hello_world/lib/main.dart
git: Branch 'beta' set up to track remote branch 'beta' from 'origin'.
Successfully switched to flutter channel 'beta'.
To ensure that you're on the latest build from this channel, run 'flutter upgrade'
scheglov@scheglov-macbookpro2:~/Source/flutter2/flutter/examples/hello_world (beta)$ /Users/scheglov/Source/flutter2/flutter/bin/flutter analyze
Downloading Dart SDK from Flutter engine 7a8f8ca02c276dce02f8dd42a44e776ac03fa9bc...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  173M  100  173M    0     0  10.0M      0  0:00:17  0:00:17 --:--:-- 10.2M
Building flutter tool...
Downloading package sky_engine...                                  469ms
Downloading flutter_patched_sdk tools...                         1,185ms
Downloading flutter_patched_sdk_product tools...                 1,235ms
Downloading darwin-x64 tools...                                     3.6s
Downloading darwin-x64/font-subset tools...                        345ms
Running "flutter pub get" in hello_world...                      1,547ms
Analyzing hello_world...

  error • The named parameter 'context' is required, but there's no corresponding argument • lib/main.dart:9:3 • missing_required_argument
   info • 'child' is deprecated and shouldn't be used. Instead of using the "child" argument, return the child from a closure provided to the "builder" argument. This
          will ensure that the BuildContext is appropriate for widgets built in the dialog. This feature was deprecated after v0.2.3. • lib/main.dart:10:5 •
          deprecated_member_use

2 issues found. (ran in 15.3s)
```

```dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/widgets.dart';
import 'package:flutter/material.dart';

void f() {
  showDialog<void>(
    child: const Text('aaa'),
  );
}

void main() =>
  runApp(
    const Center(
      child:
        Text('Hello, world!',
          key: Key('title'),
          textDirection: TextDirection.ltr
        )
      )
    );
```

**Comment by shilangyu** [comment]
Oh, alright. Sorry for the confusion.

Thanks for clarifying!
